File: core/execution/paper_broker.py
"""Paper trading broker — wraps C++ CppPaperBroker for performance."""

from __future__ import annotations

from _quant_core import CppCostModel, CppPaperBroker

from core.backtesting.cost_model import CostModel, get_cost_model
from core.execution.broker_base import BrokerBase, Order, Position
from core.execution.guard import ensure_paper_mode


class PaperBroker(BrokerBase):
    """In-memory paper trading broker with immediate fill simulation.

    Delegates order matching, position management, and P&L tracking to the
    C++ CppPaperBroker for performance. PAPER_GUARD checks stay in Python.
    """

    def __init__(self, initial_cash: float | None = None, cost_model: CostModel | None = None):
        ensure_paper_mode()
        from core.config import get_settings

        cash = initial_cash if initial_cash is not None else get_settings().PAPER_INITIAL_CASH
        cm = cost_model if cost_model is not None else get_cost_model()

        cpp_cm = CppCostModel(cm.commission_per_trade, cm.slippage_bps, cm.spread_bps)
        self._cpp = CppPaperBroker(cash, cpp_cm)
        self._cost_model = cm

    def submit_order(self, ticker: str, side: str, quantity: float, price: float) -> Order:
        ensure_paper_mode()
        side = side.upper()
        cpp_order = self._cpp.submit_order(ticker, side, quantity, price)
        return Order(
            ticker=cpp_order.ticker,
            side=cpp_order.side,
            quantity=cpp_order.quantity,
            price=cpp_order.price,
            status=cpp_order.status,
        )

    def get_positions(self) -> list[Position]:
        ensure_paper_mode()
        return [
            Position(
                ticker=p.ticker,
                quantity=p.quantity,
                avg_entry_price=p.avg_entry_price,
            )
            for p in self._cpp.get_positions()
        ]

    def get_cash(self) -> float:
        ensure_paper_mode()
        return self._cpp.get_cash()

    def get_portfolio_value(self, current_prices: dict[str, float]) -> float:
        ensure_paper_mode()
        return self._cpp.get_portfolio_value(current_prices)

    def get_orders(self) -> list[Order]:
        ensure_paper_mode()
        return [
            Order(
                ticker=o.ticker,
                side=o.side,
                quantity=o.quantity,
                price=o.price,
                status=o.status,
            )
            for o in self._cpp.get_orders()
        ]

    @property
    def realized_pnl(self) -> float:
        return self._cpp.realized_pnl()

    def unrealized_pnl(self, current_prices: dict[str, float]) -> float:
        """Compute total unrealized PnL given current market prices."""
        return self._cpp.unrealized_pnl(current_prices)

    def gross_exposure(self, current_prices: dict[str, float]) -> float:
        """Compute gross exposure as sum of absolute position market values."""
        return self._cpp.gross_exposure(current_prices)

File: cpp/src/paper_broker.cpp
#include "quant_core/paper_broker.hpp"

#include <cmath>

namespace quant_core {

PaperBroker::PaperBroker(double initial_cash, CostModel cost_model)
    : cash_(initial_cash), cost_model_(cost_model) {}

Order PaperBroker::submit_order(const std::string& ticker, const std::string& side,
                                double qty, double price) {
    double fill_price = cost_model_.apply_costs(price, qty, side);
    double commission = cost_model_.trade_commission();

    if (side == "BUY") {
        double total_cost = fill_price * qty + commission;
        if (total_cost > cash_) {
            Order order{ticker, side, qty, fill_price, "rejected"};
            orders_.push_back(order);
            return order;
        }

        cash_ -= total_cost;

        auto it = positions_.find(ticker);
        if (it != positions_.end()) {
            double old_qty = it->second.quantity;
            double old_avg = it->second.avg_entry_price;
            double new_qty = old_qty + qty;
            it->second.avg_entry_price = (old_avg * old_qty + fill_price * qty) / new_qty;
            it->second.quantity = new_qty;
        } else {
            positions_[ticker] = PosEntry{qty, fill_price};
        }

    } else if (side == "SELL") {
        auto it = positions_.find(ticker);
        if (it == positions_.end() || it->second.quantity < qty) {
            Order order{ticker, side, qty, fill_price, "rejected"};
            orders_.push_back(order);
            return order;
        }

        double proceeds = fill_price * qty - commission;
        cash_ += proceeds;

        double trade_pnl = (fill_price - it->second.avg_entry_price) * qty - commission;
        realized_pnl_ += trade_pnl;

        it->second.quantity -= qty;
        if (it->second.quantity <= 1e-9) {
            positions_.erase(it);
        }
    }

    Order order{ticker, side, qty, fill_price, "filled"};
    orders_.push_back(order);
    return order;
}

std::vector<Position> PaperBroker::get_positions() const {
    std::vector<Position> result;
    result.reserve(positions_.size());
    for (const auto& [ticker, pos] : positions_) {
        result.push_back(Position{ticker, pos.quantity, pos.avg_entry_price});
    }
    return result;
}

double PaperBroker::get_cash() const {
    return cash_;
}

double PaperBroker::get_portfolio_value(const PriceMap& prices) const {
    double value = cash_;
    for (const auto& [ticker, pos] : positions_) {
        auto it = prices.find(ticker);
        double price = (it != prices.end()) ? it->second : pos.avg_entry_price;
        value += pos.quantity * price;
    }
    return value;
}

std::vector<Order> PaperBroker::get_orders() const {
    return orders_;
}

double PaperBroker::realized_pnl() const {
    return realized_pnl_;
}

double PaperBroker::unrealized_pnl(const PriceMap& prices) const {
    double total = 0.0;
    for (const auto& [ticker, pos] : positions_) {
        auto it = prices.find(ticker);
        double price = (it != prices.end()) ? it->second : pos.avg_entry_price;
        total += (price - pos.avg_entry_price) * pos.quantity;
    }
    return total;
}

double PaperBroker::gross_exposure(const PriceMap& prices) const {
    double total = 0.0;
    for (const auto& [ticker, pos] : positions_) {
        auto it = prices.find(ticker);
        double price = (it != prices.end()) ? it->second : pos.avg_entry_price;
        total += std::abs(pos.quantity * price);
    }
    return total;
}

}  // namespace quant_core
